blueflood-core/src/test/java/com/rackspacecloud/blueflood/service/ScheduleContextAreKeysRunningTest.java

```diff
@@ -0,0 +1,212 @@
+
+        // given
+        ctx.update(currentTime - 2, shard);
+        ctx.scheduleEligibleSlots(1, 7200000);
+
+        int otherSlot = Granularity.MIN_20.slot(currentTime);
+        SlotKey otherSlotkey = SlotKey.of(Granularity.MIN_20, otherSlot, shard);
+
+        // precondition
+        assertEquals(1, ctx.getScheduledCount());
+        assertEquals(0, ctx.getRunningCount());
+
+        // when
+        boolean areKeysRunning = ctx.areChildKeysOrSelfKeyScheduledOrRunning(otherSlotkey);
+
+        // then
+        assertTrue(areKeysRunning);
+    }
+
+    @Test
+    public void unrelatedSlotScheduledReturnsFalse() {
+
+        // given
+        ctx.update(currentTime - 2, shard);
+        ctx.scheduleEligibleSlots(1, 7200000);
+
+        int otherSlot = Granularity.MIN_5.slot(currentTime - 5*60*1000); // check the previous slot from 5 minutes ago
+        SlotKey otherSlotkey = SlotKey.of(Granularity.MIN_5, otherSlot, shard);
+
+        // precondition
+        assertEquals(1, ctx.getScheduledCount());
+        assertEquals(0, ctx.getRunningCount());
+
+        // when
+        boolean areKeysRunning = ctx.areChildKeysOrSelfKeyScheduledOrRunning(otherSlotkey);
+
+        // then
+        assertFalse(areKeysRunning);
+    }
+
+    @Test
+    public void slotRunningReturnsTrue() {
+
+        // given
+        ctx.update(currentTime - 2, shard);
+        ctx.scheduleEligibleSlots(1, 7200000);
+
+        SlotKey runningSlot = ctx.getNextScheduled();
+
+        // precondition
+        assertEquals(0, ctx.getScheduledCount());
+        assertEquals(1, ctx.getRunningCount());
+        assertEquals(slotkey, runningSlot);
+
+        // when
+        boolean areKeysRunning = ctx.areChildKeysOrSelfKeyScheduledOrRunning(slotkey);
+
+        // then
+        assertTrue(areKeysRunning);
+    }
+
+    @Test
+    public void childSlotRunningReturnsTrue() {
+
+        // given
+        ctx.update(currentTime - 2, shard);
+        ctx.scheduleEligibleSlots(1, 7200000);
+
+        int otherSlot = Granularity.MIN_20.slot(currentTime);
+        SlotKey otherSlotkey = SlotKey.of(Granularity.MIN_20, otherSlot, shard);
+
+        SlotKey runningSlot = ctx.getNextScheduled();
+
+        // precondition
+        assertEquals(0, ctx.getScheduledCount());
+        assertEquals(1, ctx.getRunningCount());
+
+        // when
+        boolean areKeysRunning = ctx.areChildKeysOrSelfKeyScheduledOrRunning(otherSlotkey);
+
+        // then
+        assertTrue(areKeysRunning);
+    }
+
+    @Test
+    public void slotPushedBackReturnsTrue() {
+
+        // given
+        ctx.update(currentTime - 2, shard);
+        ctx.scheduleEligibleSlots(1, 7200000);
+
+        SlotKey runningSlot = ctx.getNextScheduled();
+
+        // precondition
+        assertEquals(0, ctx.getScheduledCount());
+        assertEquals(1, ctx.getRunningCount());
+
+        ctx.clearFromRunning(runningSlot);
+        ctx.pushBackToScheduled(runningSlot, false);
```

Comment:
@ChandraAddala It's a _possible_ scenario, although I wouldn't consider it a _valid_ one. Calling `pushBackToScheduled` while the slot is still in the running category makes it both running and scheduled. I don't believe that was the ultimate intention of the original authors, given how `getNextScheduled` is so careful to separate the two. Admittedly, this is all testing around failure conditions which I don't think the original authors adequately considered. Nevertheless, I've split it out into two separate test cases, one with and one without the call to `clearFromRunning`.